sources/Interop/D3D12MemoryAllocator/include/VirtualBlock.cs

```diff
@@ -0,0 +1,170 @@
+    {
+        VirtualBlockPimpl* m_Pimpl;
+
+        /// <summary>
+        /// Destroys this object and frees it from memory.
+        /// <para>You need to free all the allocations within this block or call <see cref="Clear"/> before destroying it.</para>
+        /// </summary>
+        public void Release()
+        {
+            using var debugGlobalMutexLock = D3D12MA_DEBUG_GLOBAL_MUTEX_LOCK();
+
+            // Copy is needed because otherwise we would call destructor and invalidate the structure with callbacks before using it to free memory.
+            ALLOCATION_CALLBACKS allocationCallbacksCopy = m_Pimpl->m_AllocationCallbacks;
+            D3D12MA_DELETE(&allocationCallbacksCopy, ref this);
+        }
+
+        /// <summary>Returns true if the block is empty - contains 0 allocations.</summary>
+        [return: NativeTypeName("BOOL")]
+        public readonly int IsEmpty()
+        {
+            using var debugGlobalMutexLock = D3D12MA_DEBUG_GLOBAL_MUTEX_LOCK();
+            return m_Pimpl->m_Metadata.IsEmpty() ? TRUE : FALSE;
+        }
+
+        /// <summary>Returns information about an allocation at given offset - its size and custom pointer.</summary>
+        public readonly void GetAllocationInfo([NativeTypeName("UINT64")] ulong offset, VIRTUAL_ALLOCATION_INFO* pInfo)
+        {
+            D3D12MA_ASSERT(offset != UINT64_MAX && pInfo != null);
+
+            using var debugGlobalMutexLock = D3D12MA_DEBUG_GLOBAL_MUTEX_LOCK();
+            m_Pimpl->m_Metadata.GetAllocationInfo(offset, pInfo);
+        }
+
+        /// <summary>Creates new allocation.</summary>
+        /// <param name="pDesc"></param>
+        /// <param name="pOffset">Offset of the new allocation, which can also be treated as an unique identifier of the allocation within this block. <see cref="UINT64_MAX"/> if allocation failed.</param>
+        /// <returns><see cref="S_OK"/> if allocation succeeded, <see cref="E_OUTOFMEMORY"/> if it failed.</returns>
+        [return: NativeTypeName("HRESULT")]
+        public int Allocate([NativeTypeName("const VIRTUAL_ALLOCATION_DESC*")] VIRTUAL_ALLOCATION_DESC* pDesc, [NativeTypeName("UINT64 *")] ulong* pOffset)
+        {
+            if ((pDesc == null) || (pOffset == null) || (pDesc->Size == 0) || !IsPow2(pDesc->Alignment))
+            {
+                D3D12MA_ASSERT(false); // "Invalid arguments passed to VirtualBlock::Allocate."
+                return E_INVALIDARG;
+            }
+
+            *pOffset = UINT64_MAX;
+
+            using var debugGlobalMutexLock = D3D12MA_DEBUG_GLOBAL_MUTEX_LOCK();
+
+            ulong alignment = pDesc->Alignment != 0 ? pDesc->Alignment : 1;
+            AllocationRequest allocRequest = default;
+
+            if (m_Pimpl->m_Metadata.CreateAllocationRequest(pDesc->Size, alignment, &allocRequest))
+            {
+                m_Pimpl->m_Metadata.Alloc(&allocRequest, pDesc->Size, pDesc->pUserData);
+                D3D12MA_HEAVY_ASSERT(m_Pimpl->m_Metadata.Validate());
```

Comment:
To continue from: https://github.com/terrafx/terrafx.interop.d3d12memoryallocator/pull/1/files#r567342315

It is very important for cases like this one, as otherwise `Validate()` is called even when asserts are "disabled".

Likewise, for `D3D12MA_HEAVY_ASSERT` it should compare against the more appropriate debug level:
```suggestion
                D3D12MA_HEAVY_ASSERT((D3D12MA_DEBUG_LEVEL > 1) && m_Pimpl->m_Metadata.Validate());
```

Reply:
Right, yeah I get it now, that's super neat 😄
This was fixed in https://github.com/terrafx/terrafx.interop.d3d12memoryallocator/commit/2acd09e47ffaf58cf960801a5f06081161a44b7e as well.